### PR TITLE
Started docstring and test coverage on Tag class

### PR DESCRIPTION
**Pull request by sveetch**
# Pull Request Check List

re: #168

- [x] Added **tests** for changed code.
- [x] Updated **documentation** for changed code.

This is and attempt to improve quality. I've started it from beast_mode as a draft. I made some tests to understand the project code and it leaded to documentation and reference tests so i would be able to know how to contribute to the new parser refactoring.

### CODE REVIEW

**Comment by christopherpickering** [comment]
@sveetch nice, thanks! What purpose does the makefile serve? Faster for you to get the project installed? I would personally add it to the gitignore.

The fixtures are looking good too. I'm still now completely satisfied with the html statements split into a list of strings, with the big print statement following, but we can revisit it.

Are you continuing to work this pr, or can I start merging it in?

Thanks!

**Comment by sveetch** [comment]
@christopherpickering yeah the Makefile is an easy entrypoint to install and performing some tasks around project. I tend to think this is commonly easy for people to quickly manipulate a project, instead of learning poetry or to learn other commands that poetry can not manage.

Whatever, i will remove it if you are not confident with it.

About the stage of this PR, it is not totally finished, i had some few method to cover for Tag, so don't merge it now, i will see when i can finish it this week, i will get back to you on discord to talk about all of this soon.

**Comment by sveetch** [comment]
@christopherpickering Ok i've finished the docstring and test coverage on Tag class.

Note than:

* The tests i've done is in "reference_tests" directory since they are totally different of your ones from tests/ dir. In the end it should be merged in test/ dir but for now they should live appart;
* There is many of "TODO" line which are notices about some incongruities and weird behaviors i found during coverage;
* poetry.lock and pyproject.toml have been automatically updated by poetry because of the development environment enabled, i don't know enough poetry to avoid this, whatever you can omit these file changes;

File: src/djlint/formatter/utils/tag.py [hunk 1]
```diff
@@ -87,14 +101,89 @@ def type(self, val: str) -> None:
...
+        TODO: Despite typed as a string, the practical usage of is_html is boolean so
+        it should be safe enough to type it as a boolean.
+
+        Arguments:
+            val (string): Value to set.
+        """
         self._is_html = val
 
+    def _spaceless_left(self) -> str:
+        """
```
Comment (christopherpickering) [inline src/djlint/formatter/utils/tag.py:135]: I recently learned from another issue that jinja also has a "+" option too. But that can be implemented later.

File: src/djlint/formatter/utils/tag.py [hunk 2]
```diff
@@ -157,47 +280,123 @@ def __get_tag_style(self, style: str) -> Dict:
...
 
     @property
     def is_indentation_sensitive(self) -> bool:
+        """
+        Check if tag is indentation sensitive.
+
+        Indentation sensitivity is based on element CSS flow. If element is HTML and
+        have a 'pre' alike flow it is considered as sensitive.
+
+        TODO: It seems '__tag_is_pre' have already been used in 'self.is_pre' from
```
Comment (christopherpickering) [inline src/djlint/formatter/utils/tag.py:332]: I agree

File: src/djlint/formatter/utils/tag.py [hunk 3]
```diff
@@ -157,47 +280,123 @@ def __get_tag_style(self, style: str) -> Dict:
...
+        Describe if Tag is an hidden element.
+
+        Element hidden status is determined from element CSS flow (block,
+        inline, table, etc..) as defined in HTML specifications, element CSS style rules
+        is not involved here.
+
+        Whatever hidden status is, if ``Tag._is_html`` is ``False`` this will always
+        return ``False``.
+
+        TODO: This property is a little disturbing with the ``Tag.hidden`` attribute
```
Comment (christopherpickering) [inline src/djlint/formatter/utils/tag.py:356]: we can prob rename this to "is_display_none" or something like that, to better represent what it is looking up in the style.

File: src/djlint/formatter/utils/tag.py [hunk 4]
```diff
@@ -157,47 +280,123 @@ def __get_tag_style(self, style: str) -> Dict:
...
     def is_script(self) -> bool:
+        """
+        Describe if Tag is assumed as a 'script' element.
+
+        Element script status is determined from the tag name.
+
+        Whatever script status is, if ``Tag._is_html`` is ``False`` this will always
+        return ``False``.
+
+        TODO: <template> element should probably be assumed as a script element.
```
Comment (christopherpickering) [inline src/djlint/formatter/utils/tag.py:380]: I would prob keep template separate. The goal of this prop is to allow the contents to be formatted w/ a js formatter, and prevent djlint from touching the contents.

I think the contents of a template tag we are generally safe to format.

**Review by christopherpickering — APPROVED** [review]

**Pull request merged**